hub-extension: Restart server with iframe of spawn page

rschroll opened the pull request:
Instead of opening the hub spawn page in a new tab, which then will redirect to open a new workspace, create an iframe for the spawn page and then close it when it redirects to the lab url.  This will maintain most of the state during restarts, although kernels will of course be restarted and any open terminals will be closed.

<del>Two</del> <ins>Three</ins> important caveats:
1. I haven't tested this in full context, since setting up a JupyterHub running a branch of JupyterLab seems fairly non-trivial.  Instead, I ran testing via an [extension](https://github.com/thedataincubator/hub-restart), which I installed in an existing JupyterHub deploy.  Once it was working, I copied the code over.
2. The current code uses a naive check to see if the spawn page had redirected back to a JupyerLab page: it simply checks if the URL matches `/lab/`.  This will obviously trigger false positives if the domain or user contains the string "lab".  I suspect there could also be false negatives, depending on options for default URLs.  I suspect the answer is to construct an expected URL from `paths.urls`, but I couldn't find the documentation about how to do that.
3. <ins>There is no meaningful error handling at the moment.  (Thanks @krassowski!)</ins>

## References

This is one solution to #17524.  The other solution mooted there (navigate to the spawn page in the current tab) did work, but causes the loss of state in any unsaved document.

<!-- Note issue numbers this pull request addresses (should be at least one, see contributing guidelines above). -->

<!-- Note any other pull requests that address this issue and how this pull request is different. -->

## Code changes

The "hub:restart" command is repurposed to open a dialog containing an iframe containing the spawn page.  We watch the location of that iframe and close the dialog once it looks like a JupyterLab URL, as that indicates that the spawn process is over.
<!-- Describe the code changes and how they address the issue. -->

## User-facing changes

After clicking "Restart" on the connection-lost dialog, users will now see the aforementioned dialog.
<!-- Describe any visual or user interaction changes and how they address the issue. -->

<!-- For visual changes, include before and after screenshots or GIF/mp4/other video demo here. -->

## Backwards-incompatible changes

None that I am aware of.

<!-- Describe any backwards-incompatible changes to JupyterLab public APIs. -->

rschroll committed:
hub-extension: Restart server with iframe of spawn page

Instead of opening the hub spawn page in a new tab, which then will
redirect to open a new workspace, create an iframe for the spawn
page and then close it when it redirects to the lab url.  This will
maintain most of the state during restarts, although kernels will
of course be restarted and any open terminals will be closed.

Right now, this uses a naive check for the redirection to the lab
URL.  We should build a better check based on path.urls.

krassowski commented:
> Does anyone know offhand what the symptoms of iframe embedding of the spawn page being blocked would be?

It looks like there is no clean solution, but timeouts can be used:
- https://github.com/whatwg/html/issues/125

We could have something like "if it does not load in 10 seconds, let's try to open in new window and close the dialog (or show another message)". What do you think?

rschroll committed:
Package integrity updates

rschroll committed:
Prettier fixes

rschroll committed:
hub-extension: Update tests to expect spawn page in iframe

Instead of spying on window.open, we inspect the DOM.  We wait for
the dialog to open, then find the first dialog element, the first
iframe inside of that, and then pull its src attribute.  This is
encapsulated in the getDialogIframeURL function.

rschroll committed:
Package integrity updates

rschroll committed:
hub-extension: Await command promises in tests

This is only needed to pass linting, but it's probably a good idea
to be sure the promises do actually resolve.

rschroll committed:
hub-extension: Fall back on iframe error

If the iframe enclosure of the spawn page raises an error for some
reason, open the spawn page in a new window, as before.  This will
catch the problem when the spawn page can't be loaded due to headers
blocking embedding, as a DOMException is raised when we try to read
iframe location.href.

rschroll commented:
It turns out that we do get a navigation event when we try to load a page blocked by `X-Frame-Options`, and then we get a DOMException when trying to read the `location.href`.  So the existing error handling should notice this problem.

I've updated things to call `window.open` with the `restartUrl` if things seem to be going wrong.  In manual testing, it seems to work okay.  I've updated the tests to notice the iframe navigation, and added one to check that it falls back to the window navigation when there's an error.  (I'm new at jest, so there may be cleaner ways to do this -- let me know if so!)

I think I've cleared up the integrity and style tests, but I had already thought that once before.

At this point, the main thing I'm worried about is checking the navigation of the iframe back to the lab URL.  If anyone knows the canonical way to check this, please let me know.

rschroll committed:
hub-extension: Better search for redirect back to Jupyterlab

This assumes the path starts /users/<username>/<paths.urls.app>,
which it does in the one case I tested.  The path could be longer,
if a particular workspace is being loaded, so don't test the full
path.

rschroll committed:
hub-extension: Fix tests

The test harness is responsible for setting the paths.  Since we're
now using paths.urls.app, and trusting it to be defined, we need to
include it in the test harness.

krassowski commented:
> I'm not sure why these tests are failing. They don't seem to be related to anything this PR was touching. Can someone point me in the right direction?


There is one UI test failing which is unrelated to this PR.

rschroll commented:
Is there anything more I need to do at this point?

krassowski commented:
It looks like there is a conflict which cannot be resolved from the GitHub UI, we cannot merge it until it the conflicts are resolved:

<img width="851" height="147" alt="image" src="https://github.com/user-attachments/assets/8298d2cf-8447-4332-8574-d890c2e08ac4" />

rschroll committed:
Merge branch 'main' into hub-restart-iframe

krassowski commented on packages/hub-extension/src/index.ts:
```diff
@@ -78,8 +80,45 @@ function activateHubExtension(
+      iframe.addEventListener('load', onIframeNavigate);
+
+      dialog = new Dialog({
+        title: trans.__('Restarting Server...'),
```
Let's use ellipsis character as it is used elsewhere

```suggestion
        title: trans.__('Restarting Server…'),
```

krassowski reviewed (commented):
Apologies that this fell through the cracks. It still seems to be in a non-mergable state:

<img width="856" height="221" alt="image" src="https://github.com/user-attachments/assets/ac6fb7d4-3077-4696-acc9-1fdf30660880" />

Would you mind creating a new branch from latest state of the `main`, applying the changes and opening a new PR?

krassowski commented:
@rschroll are you interested in finishing this up? It appears you did not check "Allow maintainers to push to this branch" so I cannot apply the suggestions; if you are no longer interested we can close this and open a new PR.
